File: 2periodo/TrabalhosPraticos/2024-1/TP1-Aeds2/Palindromo/PalindromoRecu.java
/*
*   806454 - Yago Almeida Melo
*   TP1 - Palindromo Recursivo em Java - 2024/1
*/

public class PalindromoRecu {
  /*
   * function: isFim
   * 
   * @params: (String)
   * Action: Recebe uma String e retorna verdadeiro se a String
   * é igual a "FIM", e falso se não for.
   */
  public static boolean isFim(String x) {
    return x.length() == 3 && x.charAt(0) == 'F' && x.charAt(1) == 'I' && x.charAt(2) == 'M';
  }

  /*
   * function: isPalindromo
   * 
   * @params: (String)
   * Action: Recebe uma String e retorna verdadeiro se a String
   * é um palíndromo, e falso se não for.
   */
  public static boolean isPalindromo(String x) {
    return isPalindromo(x, 0);
  }

  /*
   * function: isPalindromo
   * 
   * @params: (String, int)
   * Action: Recebe uma String e um inteiro, analisa a palavra recursivamente
   * se é ou não palindromo.
   */
  public static boolean isPalindromo(String input, int index) {
    boolean palindromo = true;
    if (index < input.length() / 2 && palindromo) {
      if (input.charAt(index) != input.charAt(input.length() - 1 - index)) {
        palindromo = false;
      } else {
        isPalindromo(input, index + 1);
      }
    }
    return palindromo;
  }

  // Main
  public static void main(String[] args) {
    String entrada = "";
    entrada = MyIO.readLine();
    while (!isFim(entrada)) {
      if (isPalindromo(entrada)) {
        MyIO.println("SIM");
      } else {
        MyIO.println("NAO");
      }
      entrada = MyIO.readLine();
    }
  }
}

File: 2periodo/TrabalhosPraticos/2024-1/TP1-Aeds2/PalindromoC/palindromoRecu.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

#define MAXLEN 1000 // Constante para definir tamanho maximo do input

bool isFim(char input[]);
bool isPalindromoRecu(char ptr[], int index);
bool isPalindromo(char input[]);

/*
* function: isFim
* @params: char []
* action: Recebe um array de char e retorna verdadeiro for igual a "FIM", senão retorna falso.
*/
bool isFim(char input[]){
	return strlen(input)==3 && input[0] == 'F' && input[1] == 'I' && input[2] == 'M';
}
/*
* function: isPalindromo
* @params: char []
* action: Recebe um array de char e retorna verdadeiro se for palindromo, senão retorna falso.
*/
bool isPalindromo(char input[]){
	return isPalindromoRecu(input, 0);
}
/*
* function: isPalindromoRecu
* @params: char [], int 
* action: Recebe um array de char e analisa se é palindromo, recursivamente, senão retorna falso.
*/
bool isPalindromoRecu(char ptr[], int index){
	bool palindromo = true;
	if(index < strlen(ptr) && palindromo){
		if(ptr[index] != ptr[strlen(ptr)-index-1]){
			palindromo = false;
		} else{
			palindromo = isPalindromoRecu(ptr, index+1);
		}
	}
	return palindromo;
}
//MAIN
int main(){
	char str[MAXLEN];
	fgets(str,sizeof(str),stdin);
	str[strcspn(str,"\n")] = '\0';  //Retira o \n e substitui por \0
	while(!isFim(str)){
		if(isPalindromo(str)) puts("SIM");
		else puts("NAO");
		fgets(str,sizeof(str),stdin);
		str[strcspn(str,"\n")] = '\0';
	}
	return 0;
}
